File: Week 06/id_501/week06/UnionFind.java
package homework.week06;

/**
 * 并查集代码模版
 * @author tangxy
 * @date 2019-11-24
 */
public class UnionFind {
    private int count = 0;
    private int[] parent;

    public UnionFind(int n){
        count = n;
        parent = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }
    }

    public int find(int p){
        while(p != parent[p]){
            parent[p] = parent[parent[p]];
            p = parent[p];
        }
        return p;
    }

    public void union(int p,int q){
        int rootP = find(p);
        int rootQ = find(q);
        if(rootP == rootQ){
            return;
        }
        parent[rootP] = rootQ;
        count --;
    }
}

File: Week 06/id_071/UnionFind.go
package week06

type UnionFind struct {
	count int
	parent []int
}

func (this *UnionFind) UnionFind(n int) {
	this.count = n
	for i := 0; i < n; i++ {
		this.parent = append(this.parent, i)
	}
}

func (this *UnionFind) find (p int) int {
	for p != this.parent[p] {
		this.parent[p] = this.parent[this.parent[p]]
		p = this.parent[p]
	}
	return p
}

func (this *UnionFind) union(p, q int){
	rootp := this.find(p)
	rootq := this.find(q)
	if rootp != rootq {
		this.parent[rootp] = rootq
		this.count--
	}
}
